### PR TITLE
[WIP] [UI] Add voice dialog

### PR DESCRIPTION
**Pull request by GiviMAD**
This is a WIP PR for the issue https://github.com/openhab/openhab-webui/issues/2275

It is built on top of the webpack 5 upgrade PR https://github.com/openhab/openhab-webui/pull/2267, because on webpack 4 you need to use loaders and I didn't want to generate more work to do in a future upgrade. The actual PR changes are in the last commit.

The PR uses these audio WebSocket which is still in review https://github.com/openhab/openhab-core/pull/4032 so you need to use the org.openhab.core.audio and org.openhab.core.voice bundles from there if you want to test it.

# Changes added:

Screenshot 2024-01-26 at 17 12 13

I have added that panel at the end of the "help & about" page (it should only be displayed if the necessary APIs are available), it contains the following options:

* "Enable Voice Dialog": When enabled the dialog execution will start after the first user event on the page (needed to start the microphone stream).
* "Connection ID": And auto generated code that is used to register the audio components. Ex: the id of the registered sink will look like "pcm::ui-52-68::sink".
* "Selected Keyword": Rustpotter config, choose rustpotter model file. At the moment the file is selected using an input type file and persisted using an IndexedDB instance.
* "Keyword Threshold": Rustpotter config, threshold in range 0 - 1.
* "Keyword Min Detections": Rustpotter config,  min number of positive scores to emit the detection.
* "Listening Item": Dialog configuration, Item to toggle while listening the command.
* "Location Item": Dialog configuration, used by the Standard language processor to improve the Item selection.

Modifying any of the options triggers a page reload.

The dialog will use the voice services configured at "Settings / Voice". Once the dialog is running it can be triggered using the voice if the rustpotter model was provided, or using the listenAndAnswer action. For example from the oh console by running `voice listenandanswer --source pcm::ui-52-68::source --sink pcm::ui-52-68::sink`.

For now I have placed some console warning in the events emitted about the dialog state. 

I wanted to open the PR in case you can give me some feedback about the integration of the feature or in case someone wants to dig into the implementation.

Best regards.

### CODE REVIEW

**Comment by GiviMAD** [comment]
Hello @florian-h05.

After the core PR has been merged I've updated the branch to the recent web-ui changes.
Everything seems to work correctly also I managed to fix some small glitches on the audio playback that I had before and wasn't able to find the cause.

I have added some descriptions to the main comment about the current state of the PR. Do you think the PR is ok for a first version? Let me know want you think when you have a moment.

In case you want to test it:
*  I downloaded the server snapshot and installed the whisper and  piper add-ons.
* For whisper, I downloaded from here https://huggingface.co/ggerganov/whisper.cpp/tree/main the [ggml-large-v3-turbo-q5_0.bin](https://huggingface.co/ggerganov/whisper.cpp/blob/main/ggml-large-v3-turbo-q5_0.bin) model and placed in the whisper folder under userdata, and then you need to select that model using the UI in the Whisper configuration.
* For piper I downloaded a voice from here https://huggingface.co/rhasspy/piper-voices/tree/main and placed it on the piper folder under userdata (both file names need to start with the same, in my case es_ES-sharvard-medium.onnx and es_ES-sharvard-medium.onnx.json).
* Then in the voice setting I set them as Default Speech-to-text and Text-to-speech services and I changed the Default Human Language Interpreter to the build in one, so it answer to you commands.
* Then launch the WebUI in dev mode and enable the dialog option and it should work.

**Comment by digitaldan** [comment]
@GiviMAD excited to try this out as well, i'll give it a go today.

**Comment by GiviMAD** [comment]
> @GiviMAD excited to try this out as well, i'll give it a go today.

Nice to hear that it will be good to have it tested on different devices.

For testing, once the connection is done, the openHAB cli can be used to record an audio and play it:
```sh
openhab> openhab:audio sinks
* PCM Audio WebSocket (ui-77-80) (pcm::ui-77-80::sink)
  System Speaker (enhancedjavasound)
  Web Audio (webaudio)
openhab> openhab:audio sources
* PCM Audio WebSocket (ui-77-80) (pcm::ui-77-80::source)
  System Microphone (javasound)
openhab> openhab:audio record pcm::ui-77-80::source 5 test_audio.wav
Recording completed
openhab> openhab:audio play pcm::ui-77-80::sink test_audio.wav
```

**Comment by GiviMAD** [comment]
Also for using it over http in other domains than localhost you need to go to `chrome://flags/#unsafely-treat-insecure-origin-as-secure` and add the openHAB server url there. If not, there is not access to the media devices and the options are hidden. I forgot to add it on the first comment.

**Comment by digitaldan** [comment]
Ok, i have upgraded my build of openHAB with the latest nightly docker image which i think has the right parts:

```
openhab> list -s|grep websock
196 │ Active │  80 │ 5.1.0.202511300254    │ org.openhab.core.io.websocket
197 │ Active │  80 │ 5.1.0.202511300258    │ org.openhab.core.io.websocket.audio
```
I have whisper and piper configured.  I also have this turned on in the Main UI (which i updated with this PR)

image

I'm connecting to OH using SSL and a real lets encrypt cert, but the mic is disabled

image

and i don't think i see anything in the logs.  Not sure if i'm missing a step? I'll try debugging a bit my self tonight and tomorrow when i have a free moment, but maybe there's something obvious @GiviMAD you can suggest?

**Comment by GiviMAD** [comment]
Yes, I tested it last night over https and it seems to  be a problem with the content security policy, I think because of the way the worklet code is packaged, I'll try to solve it in the afternoon.

**Comment by GiviMAD** [comment]
The problem with the content security policy seems to be solved by the last commit, @digitaldan let me know if it works for you now.

**Comment by digitaldan** [comment]
Excellent, i just deployed and did a real quick test an now audio is working 👍   I'll  spend some time this afternoon with it.  Thanks!

**Comment by digitaldan** [comment]
@GiviMAD excellent work, its been working great all morning, can't wait to play with this more. 

I think its about time to start looking at building a "real" AI/LLM human language interpreter for our back end that can act as a Alexa/Siri/Google replacement.  We have a ChatGPT binding, but its HLI functionality is very limited.

I was planning on writing this a year or more ago, but got wrapped up in Matter and decided that was a better use of my time for openHAB. The other, maybe larger deterrent was that all the end-to-end voice plumbing was  very overwhelming to tackle at the time......... which is exactly what you have done, and that was really the hardest part, bravo !

**Comment by GiviMAD** [comment]
> @GiviMAD excellent work, its been working great all morning, can't wait to play with this more.

Thank you for giving it a try, nice to know it's working correctly. 

One question, do you think it could be better to only load the worker and connect to the websocket by clicking in the disabled mic icon instead of doing it on the first user event in any part of the screen as is done right know? Because I did it this way but I think now it is useless to setup all of that if you are not going to use it. Maybe I should change the behavior or allow to switch it from the options.


> I think its about time to start looking at building a "real" AI/LLM human language interpreter for our back end that can act as a Alexa/Siri/Google replacement. We have a ChatGPT binding, but its HLI functionality is very limited.

There is a issue created for adding a chat to the WebUI. 
I was thinking about it, we can add a class in the server called Conversation that hold a history of messages, and we can pass this conversation as context to the interpreter as it's done right know with the location item, that way the AI interpreters (as the one in the ChatGPT binding) can feed that history to the LLM/LLM-API in order to have a conversation with them. I think this conversation can be also used as the "backend" of the chat for the UI, adding and API that allows to retrieve the conversation history and a new parameter to the interpret API that allows to send the conversation id with the new message to avoid sending the entire conversation from the UI. Also this way the chat and the voice dialog can be part of the same conversation.

I hope I have explained it correctly. I think it could be a good start point.

> bravo !

Thank you!

**Comment by digitaldan** [comment]
> do you think it could be better to only load the worker and connect to the websocket by clicking in the disabled mic icon instead of doing it on the first user event in any part of the screen as is done right know?

I think at a minimum, if the first thing the user does is click on the mic, it should connect and then actually activate the mic, right now i think you have to click twice if its the first thing interacted with?  (once to activate, then again to start the stream)

>I hope I have explained it correctly. I think it could be a good start point.

So i think you have definitely touched on part of this, and to be fair i have not looked at the interpreter framework to really understand whats built in right now, but i was thinking this LLM HLI would support:

1. Generic LLM interaction (like the chatGPT binding, so "Whats the capital of France?", "Top 10 John Hughes Movies?")
2. Tool Calling (or the equivalent functionality with whatever framework we use) 
    1. Item Control (with semantic support , synonyms, room location awareness, etc..  )
    1. Item reporting ("Whats the temperature in the Kitchen?")
    1. Rule control, execution and even the ability to create rules and maybe one time rules (like "turn the lights off in 10 mins") 
3. Memory, history and awareness of previous interactions
4. LLM agnostic, can use open ai, anthropic, gemini .....or even better, local models through ollama and other services which support  the common API many local inference servers use.   
5. Probably other things i am not remembering right now.....

I need to spend some time to really think about it, but again, you solved the hardest part in my mind, while chat is neat and a great way to test (and i could see hooking up to SMS or slack for chatting with your home while away), voice is really the killer application for this.

**Comment by GiviMAD** [comment]
> So i think you have definitely touched on part of this, and to be fair i have not looked at the interpreter framework to really understand whats built in right now, but i was thinking this LLM HLI would support:
> 
>     1. Generic LLM interaction (like the chatGPT binding, so "Whats the capital of France?", "Top 10 John Hughes Movies?")
> 
>     2. Tool Calling (or the equivalent functionality with whatever framework we use)
>        
>        1. Item Control (with semantic support , synonyms, room location awareness, etc..  )
>        2. Item reporting ("Whats the temperature in the Kitchen?")
>        3. Rule control, execution and even the ability to create rules and maybe one time rules (like "turn the lights off in 10 mins")
> 
>     3. Memory, history and awareness of previous interactions
> 
>     4. LLM agnostic, can use open ai, anthropic, gemini .....or even better, local models through ollama and other services which support  the common API many local inference servers use.
> 
>     5. Probably other things i am not remembering right now.....

I have a similar vision but I think not exactly the same. What I think is that most of those capabilities should be encapsulated in the core (history, tools, memory....) and the HLI interface keeps been a black box that allows to connect to whatever implementation you want (OpenAI, Ollama, Gemini... or the current built-in interpreter). I haven't got much experience with this but I was looking around and that seems to be a good way to go and I think is not hard going from the current state to that.

> I think at a minimum, if the first thing the user does is click on the mic, it should connect and then actually activate the mic, right now i think you have to click twice if its the first thing interacted with? (once to activate, then again to start the stream)

Right now, once you enable it, a click on any place of the page loads the Webworker and connects to the Websocket.
About triggering it on the first click I see it can be useful depending on how you use it, but some times I want to launch it and not use it in the moment, I just want the sink and source to be registered on the server. And in case of adding a keyword spotter (which I hope I can add in the future) it makes sense to have the setup and the trigger in different stages.

I think I will add those two things as options and move the PR to ready, thank you for the feedback.

**Comment by florian-h05** [comment]
Can't wait to try this out, but unfortunately I am quite busy at the moment ...

Just wanted to share a few issues/PRs with you wrt HLI:

- https://github.com/openhab/openhab-webui/issues/2995
- https://github.com/openhab/openhab-core/pull/5016
- https://github.com/openhab/openhab-addons/pull/19267

**Comment by GiviMAD** [comment]
> Can't wait to try this out, but unfortunately I am quite busy at the moment ...
> 
> Just wanted to share a few issues/PRs with you wrt HLI:
> 
>     * [Add a intelligent smart home chat bot to the UI #2995](https://github.com/openhab/openhab-webui/issues/2995)
> 
>     * [Add HLI bundle with EnhancedHLIInterpreter and shared Card/Component model openhab-core#5016](https://github.com/openhab/openhab-core/pull/5016)
> 
>     * [[ChatGPT] enhanced HLI service openhab-addons#19267](https://github.com/openhab/openhab-addons/pull/19267)

Thank you for the links @florian-h05 

I have taken a quick look and I don't think the best choice is changing the interpreters response.

I think that if you want to allow the LLM model to have the ability to display cards on the chat in the UI you should expose that as a tool to the model instead of it been the model response. For example if you want to implement the capability of generating images using the chat, you will make a new tool available to the model and when it calls that tool the image will be send to the chat, it doesn't need to be integrated into the interpreter response, and also changing the interface will made then unusable by audio.

**Comment by florian-h05** [comment]
> I have taken a quick look and I don't think the best choice is changing the interpreters response.

My proposal would be to extend the response with tool calls, so a model can return both text and tool calls. That’s what you suggest as well.
I have to admit I haven’t looked at the PRs I linked yet (lack of time).

**Comment by GiviMAD** [comment]
> > I have taken a quick look and I don't think the best choice is changing the interpreters response.
> 
> My proposal would be to extend the response with tool calls, so a model can return both text and tool calls. That’s what you suggest as well. I have to admit I haven’t looked at the PRs I linked yet (lack of time).

I think that by adding a "Conversation" that the interpreter can interact with in real time without waiting for the interpretation to end makes more sense, there you can have a conversation between different roles, the basic ones like the user and openhab but also the thinking and the tool calling, and the chat can get the changes of the conversation in real time and make the user aware of the tool usage and thinking, and also implement the cool text streaming that is common on the AI chats.

I think it solves most of the problems without requiring any breaking change. I'll try to write a draft PR to see if it makes sense or I'm missing something.

**Comment by digitaldan** [comment]
> I have a similar vision but I think not exactly the same. What I think is that most of those capabilities should be encapsulated in the core (history, tools, memory....) and the HLI interface keeps been a black box that allows to connect to whatever implementation you want (OpenAI, Ollama, Gemini... or the current built-in interpreter). I haven't got much experience with this but I was looking around and that seems to be a good way to go and I think is not hard going from the current state to that.

Yes that sounds like a solid plain, i like the idea of breaking out the functionality into separate features or even bundles

> Just wanted to share a few issues/PRs with you wrt HLI:

Well,  thats always strange when you start reading a thread and stumble upon your own posts which you totally don't remember writing ;-). ....... so thanks for the reminder.  Like groundhog's days.

>I think it solves most of the problems without requiring any breaking change. I'll try to write a draft PR to see if it makes sense or I'm missing something.

That would be great, I would love to see a high level design of how this would all fit together end-to-end, i think that would really speed things up.... right now its a bit ambiguous to me (  hoping to dive into the PRs mentioned as well as the HLI code this weekend as i am totally not up to speed)

**Comment by Artur-Fedjukevits** [comment]
> I have taken a quick look and I don't think the best choice is changing the interpreters response.
> 
> I think that if you want to allow the LLM model to have the ability to display cards on the chat in the UI you should expose that as a tool to the model instead of it been the model response. For example if you want to implement the capability of generating images using the chat, you will make a new tool available to the model and when it calls that tool the image will be send to the chat, it doesn't need to be integrated into the interpreter response, and also changing the interface will made then unusable by audio.

Hi @GiviMAD! Sorry for the late reply — I only just noticed your message.

In that PR the existing HLI response is not changed.  
`HumanLanguageInterpreter#interpret()` stays exactly as it is and is still used by
the audio pipeline, so audio use cases are not affected.

`EnhancedHLIInterpreter` is an optional extension that is only used by HABot.
If an interpreter doesn’t implement it, HABot simply falls back to `interpret()`.
This keeps the solution fully backward compatible.

This PR mainly moves the existing model into core to avoid
cross-addon dependencies, not to introduce UI-specific logic.

A tool-based LLM approach could be interesting in the future, but it would require
a much larger framework (tool lifecycle, routing, sessions, etc.). This PR is meant
as a small, incremental step that works with the current HABot architecture and
doesn’t block future evolution.

In short: audio stays unchanged, existing HLIs stay unchanged, and chat gets an
optional richer response channel.

This approach follows the direction recommended by @florian-h05  during the
earlier discussions.

**Comment by florian-h05** [comment]
@GiviMAD Thanks for the PR - amazing work! 
Sorry for the super late review, I rebased yesterday to fix all conflicts and converted the code to typescript. I will also address the Copilot review once that's finished.

File: bundles/org.openhab.ui/web/src/js/voice/audio-main.ts [hunk 1]
```diff
@@ -0,0 +1,428 @@
...
+    const closeMsg = this.events.onMessage?.('Resuming audio context, click to continue', 'info')
+    document.addEventListener('click', resumeAudioContext)
+    await this.getAudioSource().resume()
+    document.removeEventListener('click', resumeAudioContext)
+    closeMsg?.()
+    document.removeEventListener('visibilitychange', this.handleSuspendOnHidden)
+    if (speakerConfig.suspendOnHide) {
+      document.addEventListener('visibilitychange', this.handleSuspendOnHidden)
+    }
+    await AudioSink.configure(audioContext, speakerConfig.useAudioElement)
```
[the remark this answers is not in the document]
Comment (florian-h05) [reply bundles/org.openhab.ui/web/src/js/voice/audio-main.ts:347]: The volume is passed upon AudioSink instance creation in line 249.

**Review by florian-h05 — APPROVED** [review]
LGTM!!

One question though: What happened to the keywordspotter support from https://github.com/openhab/openhab-webui/tree/355c5fca7fd15b5fd89bcac7039bf27b04a3273b/bundles/org.openhab.ui/web/src/js/voice?

**Comment by GiviMAD** [comment]
Thank you for the review/merge!

> One question though: What happened to the keywordspotter support from https://github.com/openhab/openhab-webui/tree/355c5fca7fd15b5fd89bcac7039bf27b04a3273b/bundles/org.openhab.ui/web/src/js/voice?

The integration worked correctly but I wanted to rebuilt rustpotter using the google feature extractor model as openwakeword do to try to achieve generic models generated with TTS samples, because with the current implementation it doesn't work, but I never got into it.

Having this merged is a great motivator to give it a try, so thank you again.

**Comment by florian-h05** [comment]
Okay, I'll then look forward to keyword spotting in the UI!
Again: Amazing work you've done and do on all the audio and voice stuff, especially the encoding/decoding, audio formats etc. is really advanced stuff to me.
